### DISCUSSION TITLE
Dynamic chart

### DISCUSSION THREAD

**Discussion by asynchaizer** [KoalaPlot/koalaplot-core#86] [Q&A] [answered]
Hi, I have a need for a line graph that can work with dynamic data that comes from a sensor at high speed (more than 50-100 points per second), the sensor can be turned off at any time and the readings obtained can be studied.
I have a number of questions:
1. Can I work with dynamic data? Are there any optimizations for this?
2. Is it possible to set pan restrictions? Here I am more interested in the width limit (Maximum number of displayed values)
3. Can I programmatically move the panoramic view to the end each time new values are added? I understand that there is no desktop solution right now, but I would like to clarify this within the Android.

**Comment by gsteckman** [comment, marked as answer]
1. Yes it can work with dynamic data. There is an example in the Samples project, have a look at LiveTimeChart.kt. Specific optimizations have not been made for performance. But practically speaking if a person is looking at the data I don't think it would require a 100 Hz refresh rate, a few times per second is likely sufficient.
2. Yes this can be set by adjusting the min/max range of values to be plotted on an axis.
3. Yes, also controlled by the min/max range on the axis. There is a desktop solution but manually zooming/panning with mouse/touch only works on Android for now (Compose limitation - maybe it works on Apple devices but I don't have any to test with so I don't know).

**Reply by asynchaizer** [reply]
I have studied some of the source code and understand that panning actually works on desktop, but there is a problem.

I believe that the implementations of the "AxisModel" interface are not flexible enough to "activate" this feature, and in fact only zoom does not work for desktop(but it seems to me that this is solved by scrolling).
Perhaps I would like to extend the functionality of AxisModel implementations by making them more flexible.:
1. Range - I would perceive it as it is now, without changes. But I would not like it to be a zone that is the current "ViewPort", I would like it to be a zone that cannot be left by panning, that is, the entire available area of the chart.
2. I suggest adding parameters for the "ViewPort" that would allow you to limit the minimum and maximum viewport. Sometimes it is pointless to give the opportunity to view all the values of the chart. For example, we have 100,000+ points (low density) - these readings cannot be considered usefully, compression algorithms (such as LTTB) are needed for this, and even with such optimization it does not always make sense. I see that there is a "zoomRangeLimit", perhaps this is the minimum "ViewPort".
3. I suggest adding the "initialViewport" feature, this is the section of the chart where the "ViewPort" will be moved when the chart appears, for example - I want to focus the user's attention on a problematic area on the chart. The "initialViewPort" can be "currentRange", perhaps it remains only to add an input parameter to set the value for "currentRange". Perhaps it is worth allowing it to be changed, for example, there is a list of problem areas, and I want to give the opportunity to move from one zone to another.

I tried to show it graphically, I hope it helps to understand the idea. 
What do you think about it?
![image](https://github.com/user-attachments/assets/1c3af4bf-1f7c-4265-adad-0372c69ea217)

**Reply by gsteckman** [reply]
I might have misrepresented the problem on desktop. Panning and zooming both work in the sense that the pan/zoom values can be changed on the axis and the plot will redraw to reflect that. What doesn't work on Desktop is the mouse input for zooming, because last I checked Compose Desktop (and web) didn't support the mouse ctrl+scroll wheel or other typical key/mouse combinations for zooming.

The current implementation has all the flexibility to do what you are describing (except downsampling), it's just represented numerically in a different way. The way the zoom and pan functions currently work is aligned with how the mouse/gestures come in from compose, e.g. see `PointerInputScope.detectTransformGestures`. But the zoom/pan functions transform their inputs to a min/max range, stored internally in `currentRange` (e.g., in `FloatLinearAxisModel`). I think this is equivalent to what you are calling ViewPort.

`zoomRangeLimit` is the maximum amount allowed to zoom in on an axis - if inputs try to exceed this they will be ignored. Another way to say it is as a minimum viewport size, in that it is the minimum range between max/min values that are allowed to be displayed on the axis.

What would be pretty easy to do here is expose `currentRange` through a new function on AxisModel called `setCurrentRange()` with an empty implementation (some axes do not currently allow zooming, eg. `CategoryAxisModel` and `LogAxisModel`). The various LinearAxisModels can put some input bounds checking on the function and then just set their `currentRange` variable. This may be an easier way to programmatically change the zoom of an axis when not passing inputs directly from the Compose mouse/gesture handler. It's a direct way to set the values, whereas the zoom/pan approach is always relative to the current settings (since it follows the mouse input approach which is a stream of continuous delta updates).

**Reply by gsteckman** [reply]
I just pushed a commit with this new function to main, but it's still untested.

**Reply by asynchaizer** [reply]
Thanks for the detailed explanations! Do I understand correctly that it is necessary to add tests for this function? (I could probably contribute accordingly)

**Reply by asynchaizer** [reply]
Another point that is related to the "currentRange", now there is no direct way to limit its maximum allowable range? The "range" parameter, which exists, limits the entire area of the graph that the user can see, but what about the maximum field of view? Conditionally, we have data N for 10 years with an accuracy of up to an hour, how to limit the possibility of increasing the viewing area by "X" so that the user does not see all the readings for 10 years, but only for 1 month? (At the same time, make it possible to move along the schedule by panning along the "X" axis within 10 years)

**Reply by gsteckman** [reply]
On the tests, I was just warning that I haven't tested it, end even though it looks simple maybe it has a bug and won't work. If you'd like to add a test for it that would be great and much appreciated.

For the ranges, I think the way it could work is to do something like:

```
val range = LongLinearAxisModel(0..87600, zoomRangeLimit = 700)
range.setViewRange(0..720) // start by viewing the first month of data

... pass it to the XYGraph
```

Then if the user scrolls, however you get that input, you can update setViewRange, e.g. if you want to view data starting from day 10 `range.setViewRange(240..960)`.

You can put the LongLinearAxisModel in a MutableState holder so the plot recomposes when you update the range.

**Reply by asynchaizer** [reply]
"zoomRangeLimit" limits the minimum viewing area, not the maximum. Shouldn't we introduce a new parameter that would limit the maximum viewing area?

After testing the zoom gestures, I realize that it would be more convenient to adjust the zoom independently for each axis. So far, I see that this is a problem in Compose, namely in "detectTransformGestures" - the "zoom" parameter provides a general quadratic scaling, rather than separate scaling along the x and y axes.

I have attached an example of the operation of the relatively old library "MPAAndroidChart", the example shows how the maximum viewing area is limited along the X axis, and how the scaling along the axes is controlled separately.

https://github.com/user-attachments/assets/05d06843-5b3d-44fc-8746-01e2bf06de85

**Reply by gsteckman** [reply]
`range` will limit the amount you can "zoom out", and `zoomRangeLimit` limits the amount you can "zoom in".

As far as zooming each axis separately or by different amounts from user input/gestures, I am not sure how that can be done if Compose only provides a single zoom value and not separate values for each axis. But since the axis models have independent zoom settings, it is possible to do it programmatically.

**Reply by asynchaizer** [reply]
That's right, zoom limits the maximum "zoom out", but at the same time range is also the maximum limit for the viewing area.
I would like to make it possible to pan through the diagram, but at the same time limit the zoom out.

For example, let's take a chart with an X-axis range from 0 to 10_000, I want to limit the "zoom in" to 10 and the "zoom out" to 100. If the user zoomed out to 100, then he can view the values on the left or right only by panning. I will not be able to do this through range, because if I set "0..100" for the range value, then when zooming out, I will not be able to pan to the rest of the diagram, because range will also limit panning.

**Reply by gsteckman** [reply]
OK I think I know what you mean. Are you suggesting an upper limit to the `currentRange` value that will be enforced by the `zoom` and `setViewRange` functions?

**Reply by asynchaizer** [reply]
Yes, you're right. Probably "zoomRangeLimit" should be replaced with "zoomInLimit", and also add "zoomOutLimit". Alternatively, for "zoomRangeLimit", replace the parameter type with Range, but in this case there may be a dual understanding - "min is zoomIn or zoomOut?", so I would prefer more explicit names that directly reflect the essence.

**Reply by gsteckman** [reply]
I think keeping with the data type of the axis and having 2 values is going to be more clear than using a Range. Other name options are minViewExtent/maxViewExtent. I'm not sure if that is more clear than zoomInLimit/zoomOutLimit?

**Reply by asynchaizer** [reply]
The names "zoomInLimit/zoomOutLimit" are directly related to zoom, they are intuitive and emphasize the scaling process. The names "minViewExtent/maxViewExtent" - I have not seen names in practice that were associated with "viewExtent", but I will assume that this option is as good as "zoomInLimit/zoomOutLimit", but this option focuses on the displayed range of data, and not on the concept of scaling.
Since these names will limit not only the scaling process, but also the setting of the current range (setViewRange), I think the names "minViewExtent/maxViewExtent" would be preferable, since in the future a function may appear that is not related to scaling, but these parameters will affect this function.

**Reply by gsteckman** [reply]
Hi @asynchaizer, after the gesture improvements you recently made, do you think it would still be helpful to have a minViewExtent/maxViewExtent setting on the linear axis? So zoomRangeLimit would be replaced with minViewExtent, and a new parameter maxViewExtent would be introduced to potentially prevent zooming out all the way to the maximum range of the data.

**Reply by asynchaizer** [reply]
Hi @gsteckman, I still think it needs to be implemented.
This will be especially useful on large data samples, because I noticed a decrease in rendering performance in such cases, the problem is observed from about 2000 points in the visible range, in the near future I plan to familiarize myself with this problem in more detail.

**Comment by gsteckman** [comment]
Hi @asynchaizer, I just committed this change. Do you have any use cases to try it out with?  I haven't tried doing any scale testing or optimization of the library yet, but that would definitely be interesting to see if you have the time.

**Reply by asynchaizer** [reply]
Hi, @gsteckman!

I have checked and it works well. There is one problem, for `currentRange` it is necessary to apply a restriction when initializing `AxisModel`, since now the restrictions of the `viewExtent` are applied only when changing the `currentRange` through the `setViewRange` method.

There is also a loosely coupled problem in the `zoom` implementations - this is not critical, but it affects the UX charts. When the zoom is maximum or minimum, when trying to zoom in/out, only the zoom offset starts working, but in fact I expect that there should be no offset, because the offset should occur with zoom in/out.

Here's what I managed to find when reviewing performance, I tested only LinePlot on real data - 3 lines, totaling ~22700 points :
1. The highest load occurs due to the rendering of `Symbols`, performance drops to <1fps or rendering is completely blocked and the application freezes. It seems inevitable and easier not to use `Symbols` on big data. The only idea is to provide a more flexible implementation of `Symbols`, for example, so that it is possible to put dots only in the right places, for example, at peak values or only for maximum/minimum values. I think the performance issue when using `Symbols` should be documented.
2. I also noticed that `Path` is used to build lines, as far as I know `Path` is slower than native calls to `Canvas` methods, I tried to quickly rewrite it to the native `Canvas.drawLine` method and got a performance gain of ~10-15 fps. Despite the fact that the `lineTo` method inside `Path` works 2 times faster than `Canvas.drawLine`, drawing `Path` inside `Canvas` eventually turns out to be more difficult with an increase in the number of points.
3. It may be worth checking the `Canvas.scale` and `Canvas.translate` methods, perhaps they will be more productive.
4. We can potentially get a performance boost if we select only those points that are really needed for rendering each time. Now, every time we draw, the `lineTo` method is called for all points, but in fact it is enough to draw only those points that are in the visible range.
5. We can increase productivity with several lines in the diagram if we draw them in parallel, for example, now if we draw 3 lines, it means that the cycle will be started 3 times, but in fact it is possible to draw lines in 1 iteration of the cycle. This requires passing all lines to 1 LinePlot function, so it will require some API changes. If we add the previous paragraph to this point, then this way we can reach the limit of efficiency in collecting the necessary points for rendering

**Reply by gsteckman** [reply]
Good catch on those min/max view extent issues. I checked in code that should fix them.

Path was used in order to support area plots so the same initial path could be used to draw the line and then form one boundary of the fill area. There's also been some interest in being able to draw smoothed curves, and path can draw Bezier curves but as far as I know DrawScope doesn't have a function to do it.

Can you elaborate on what you mean with point 3?

I'm not following you on point 5. Since Compose is single threaded we can't launch separate threads, and I assume the loop overhead is small, which is typically the case, relative to the line calls.

Thanks for looking into this and for the input.

**Reply by asynchaizer** [reply]
Hi @gsteckman !

2 - We can make the Path a separate implementation, especially for Bezier curves. We can also implement the Bezier algorithm ourselves using DrawScope, this is definitely a more difficult way (or maybe more interesting?)

3 - I mean, perhaps the pan and zoom functions can be performed using the native DrawScope.scale and DrawScope.translate methods. But so far, these are just abstract assumptions.

5 - Yes, that's right, rendering takes place in one thread. But not everything is rendering, some of the calculations can be done in the background, for example, you can see an example of such a function [here](https://github.com/maximilianproell/compose-multiplatform-chart/blob/300f6c69b4d7a5edfeadb661f58e27fbfb0bd886/multiplatformchart/src/commonMain/kotlin/at/maximilianproell/multiplatformchart/linechart/LineChart.kt#L174).
Now such logic is difficult to implement, since it is almost impossible and very expensive to synchronize the construction of two or more lines on the chart at the same time, because now each line is processed independently of the other. 

Perhaps this is one of the ways to reduce or eliminate the impact of the complexity of the amount of data on the UI stream.
On the other hand, if we put a number of calculations in the background, we will get a delay when using real-time diagrams, which is not always acceptable 🤔

**Reply by asynchaizer** [reply]
In 5, I rather meant to start by trying to draw all the lines within 1 LinePlot, so we will have 1 Canvas for the lines of the diagram and a minimum number of accesses to data samples by drawing all the lines in 1 iteration of the cycle. I'm not ready to say how much more productive it is yet, but it should definitely save rendering time.
